sktime/utils/tests/test_datetime.py

```diff
@@ -88,3 +95,56 @@ def test_infer_freq() -> None:
     y = _bottom_hier_datagen(no_levels=2)
     y = VectorizedDF(X=y, iterate_as="Series", is_scitype="Hierarchical")
     assert infer_freq(y) == "M"
+
+
+def set_freq() -> None:
+    """Test frequency inference."""
+    assert infer_freq(None) is None
+
+    y = pd.Series(dtype=int)
+    assert infer_freq(y) is None
+
+    index = pd.date_range(start="2021-01-01", periods=1)
+    y = pd.Series(index=index, dtype=int)
+    assert infer_freq(y) == "D"
+
+    index = pd.date_range(start="2021-01-01", periods=1, freq="M")
+    y = pd.Series(index=index, dtype=int)
+    assert infer_freq(y) == "M"
+
+    y = pd.DataFrame({"a": 1}, index=pd.date_range(start="2021-01-01", periods=1))
+    assert infer_freq(y) == "D"
+
+    y = pd.DataFrame(
+        {"a": 1}, index=pd.date_range(start="2021-01-01", periods=1, freq="M")
+    )
+    assert infer_freq(y) == "M"
+
+    y = _bottom_hier_datagen(no_levels=2)
+    y = VectorizedDF(X=y, iterate_as="Series", is_scitype="Hierarchical")
+    assert infer_freq(y) == "M"
+
+
+def test_set_freq_hier():
+    """Test that setting frequency on a DatetimeIndex MultiIndex works."""
+    y = load_airline()
+
+    assert get_time_index(y).freq is not None
+
+    # Convert to DatetimeIndex
+    y.index = y.index.to_timestamp()
+
+    assert get_time_index(y).freq is not None
+
+    # Create MultiIndex
+    mi = pd.MultiIndex.from_product([[0], y.index], names=["instances", "timepoints"])
+    y_group1 = pd.DataFrame(y.values, index=mi, columns=["y"])
+
+    mi = pd.MultiIndex.from_product([[1], y.index], names=["instances", "timepoints"])
+    y_group2 = pd.DataFrame(y.values, index=mi, columns=["y"])
+
```

Comment:
I think we should first assert that it is `None` to start with, and add a comment that this is intentionally creating the problematic case?